docs/specs/captcha.md

```diff
@@ -1,41 +1,328 @@
+
+### authgear.yaml
 
 ```yaml
 captcha:
-  provider: cloudflare
+  enabled: true
+  providers:
+  - type: cloudflare
+    alias: cloudflare
 ```
 
-- `provider`
-  - The captcha provider. Currently, only `cloudflare` is supported.
+- `captcha.enabled`: Boolean. If it is true, the new configuration is used.
+- `captcha.providers`: An array of Captcha provider configuration. The actual shape depends on the `type` property.
+- `captcha.providers.type`: Required. The type of the Captcha provider. Valid values are `cloudflare` and `recaptchav2`.
+- `captcha.providers.alias`: Required. The unique identifier to the Captcha provider. It is used in other parts of the configuration to refer to this particular Captcha provider. For example, the project can configured a number of Captcha providers, but only uses one of them in a particular Authentication Flow.
 
-### authgear.secrets.yaml
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
 
-When the captcha provider is configured, the corresponding secret must exist in `authgear.secrets.yaml`.
+There is no specific fields.
+
+### authgear.secrets.yaml
 
 ```yaml
 - data:
-    secret: YOUR_TURNSTILE_SECRET_KEY
-  key: captcha.cloudflare
+    items:
+    - type: cloudflare
+      alias: cloudflare
+      secret: TURNSTILE_SECRET_KEY
+  key: captcha.providers
+```
+
+- `items.type`: Required. It is the same as `captcha.providers.type`.
+- `items.alias`: Required. It is the same as `captcha.providers.alias`.
+
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
+
+- `secret_key`: Required. The Cloudflare Turnstile secret key.
+
+## Authentication Flow
+
+This section specifies how Captcha works in a Authentication Flow.
+
+### Captcha in Authentication Flow configuration
+
+Captcha is supported in the following flow types:
+
+- `signup`
+- `promote`
+- `login`
+- `signup_login`
+- `reauth`
+- `account_recovery`
+
+Captcha is supported only in the following step types:
+
+- `identify` in `signup`, `promote`, `login`, `signup_login`, and `account_recovery`.
+- `create_authenticator` in `signup` and `promote`.
+- `authenticate` in `login` and `reauth`.
+
+We can see that all supported step types have branches.
+To specify Captcha is required, add `captcha.required=true` to the branch.
+
+For example,
+
+```
+authentication_flow:
+  signup_flows:
+  - name: default
+    steps:
+    - type: identify
+      one_of:
+      - identification: email
+        # Identify with email requires captcha.
+        captcha:
+          required: true
+    - type: authenticate
+      one_of:
+      - authentication: primary_password
+      - authentication: primary_oob_otp_email
+```
+
+### Default behavior
+
+When Captcha is enabled and there is at least one configured provider,
+all the branches of the first step (that is, the `identify` step) of the generated flows will have `captcha.required=true`.
+
+This means whether Captcha is enabled, every flow requires captcha at the beginning of the flow.
```

Comment:
If 
- more than 1 configured provider
- user did not specify which provider to use

Will
1. throw error; OR
2. randomly select a configured provider; If this, any selection principle?

Reply:
Added a commit to clarify.